src/v/hashing/secure.h

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+#include "bytes/bytes.h"
+
+#include <gnutls/crypto.h>
+#include <gnutls/gnutls.h>
+
+class hmac_exception : public std::exception {
+public:
+    explicit hmac_exception(const char* msg)
+      : _msg(msg) {}
+
+    const char* what() const noexcept final { return _msg; }
+
+private:
+    const char* _msg;
+};
+
+namespace internal {
+
+template<gnutls_mac_algorithm_t Algo, size_t DigestSize>
+class hmac {
+    static_assert(DigestSize > 0, "digest cannot be zero length");
+
+public:
+    explicit hmac(const bytes& key)
+      : hmac(reinterpret_cast<const char*>(key.data()), key.size()) {} // NOLINT
+
+    hmac(const hmac&) = delete;
+    hmac& operator=(const hmac&) = delete;
+    hmac(hmac&&) = delete;
+    hmac& operator=(hmac&&) = delete;
+
+    ~hmac() noexcept { gnutls_hmac_deinit(_handle, nullptr); }
+
+    void update(const bytes& data) {
+        // NOLINTNEXTLINE
+        update(reinterpret_cast<const char*>(data.data()), data.size());
+    }
+
+    template<std::size_t Size>
+    void update(const std::array<char, Size>& data) {
+        update(data.data(), Size);
+    }
+
+    /**
+     * Return the current output and reset.
+     */
+    std::array<char, DigestSize> reset() {
+        std::array<char, DigestSize> digest;
+        gnutls_hmac_output(_handle, digest.data());
+        return digest;
+    }
+
+private:
+    hmac(const char* key, size_t size) {
+        int ret = gnutls_hmac_init(&_handle, Algo, key, size);
+        if (unlikely(ret)) {
+            throw hmac_exception(gnutls_strerror(ret));
+        }
+
+        ret = gnutls_hmac_get_len(Algo);
+        if (unlikely(ret != DigestSize)) {
+            throw hmac_exception("invalid digest length");
+        }
+    }
+
+    void update(const char* data, size_t size) {
+        int ret = gnutls_hmac(_handle, data, size);
+        if (unlikely(ret)) {
+            throw hmac_exception(gnutls_strerror(ret));
+        }
+    }
+
+    gnutls_hmac_hd_t _handle{};
```

Comment:
interesting. often these C libs have a <foo>_init() methods. surprised gnutls doesn't.

Reply:
gnutls does and its used right above this.

Reply:
that initialization... mm was done to quiet the linter. will actually remove the `{}` since gnutls is already initializing this.